docs/source/reference/api-server/api-server-upgrade.rst

```diff
@@ -0,0 +1,240 @@
+.. _sky-api-server-upgrade:
+
+Upgrading SkyPilot API Server
+=============================
+
+This page provides an overview of the steps you should follow to upgrade a remote SkyPilot API server:
+
+* :ref:`sky-api-server-helm-upgrade`
+* :ref:`sky-api-server-vm-upgrade`
+
+.. _sky-api-server-helm-upgrade:
+
+Upgrade API server deployed with Helm
+-----------------------------------------
+
+Here we introduce the steps for upgrading a remote API server deployed with :ref:`Helm deployement <sky-api-server-deploy>`.
+
+.. note::
+
+    Upgrading the API server introduces downtime. We recommend to schedule the upgrade during a **maintenance window**: cordon & drain the old API server and upgrade.
+
+Step 1: Prepare an upgrade
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Find the version to use in SkyPilot `nightly build <https://pypi.org/project/skypilot-nightly/#history>`_.
+2. Update SkyPilot helm repository to the latest version:
+
+.. code-block:: bash
+
+    helm repo update skypilot
+
+3. Prepare variables for the upgrade, the ``NAMESPACE`` and ``RELEASE_NAME`` should be the same as the previous deployment step:
+
+.. code-block:: bash
+
+    NAMESPACE=skypilot # TODO: change to your installed namespace
+    RELEASE_NAME=skypilot # TODO: change to your installed release name
+    VERSION=1.0.0-dev20250410 # TODO: change to the version you want to upgrade to
+    IMAGE_REPO=berkeleyskypilot/skypilot-nightly
+
+Step 2: Upgrade the API server and clients
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. note::
+
+    * Upgrading the API server will interrupt any pending and on-going requests on the API server, e.g., on-goining ``sky launch``. You can recover on-going requests by running the same commands again after the new API server starts.
+    * Upgrading only one of the server or clients may break the compatibility between the server and clients.
+
+    To minimize the impact of upgrading, you can :ref:`cordon and drain the API server <cordon-drain-api-server>` before upgrading the API server and clients to enforce that all the requests to be finished before the upgrade and there is no new request during the upgrade.
+
+Upgrade the clients:
```

Comment:
When reading here, an admin may think "how can I make sure all the clients (humans) run this"? Which leads to the question: should we add a brief section on client-server back-compat guarantees? E.g., admin may want to perform the server upgrade, and have human clients asynchronously upgrade their client versions.

Reply:
We can have this after https://github.com/skypilot-org/skypilot/issues/5184

The first version of this doc asked users to ensure all the clients used by pipeline are upgraded before upgrading the server and the clients used by human developers can be upgraded asynchronously when the user hit the API error  in the next time. This was simplified to current one to keep brevity, how do you like?

Reply:
If clients (humans) didn't do this and they send a request to an updated API server, what happens? Is there a nice version mismatch error? I feel this is important to hint about in this doc.

Reply:
Good point, making some refinement